### ISSUE TITLE
What is the strategy of Celery get task messages and execute?

### ISSUE THREAD

**Issue by jcyrss** [celery/celery#3394]
I've learned from Celery doc, by default, Celery will acknowledge message before it executing the task.

Will a Celery worker process retrieve mutiple task messages, and ack them together, or it just retrieve one, ack one and execute one, then retrieve next one, ack it and execute it?

If it is the latter, for a worker process in greenlet mode, when it is executing a task and waiting for a network response, will it start new greenlet to retrieve another task message and execute it, or it just wait for the first task to complete?

**Comment by ask** [comment]
There's one greenlet thread responsible for receiving messages (task consumer), and there's a pool of greenlet worker threads.

The task consumer receives one task message and sends it to the pool, then receives another task message, sends it to the pool, and so on.

A pool worker thread receives the task message and acknowledges that it starts working on it by sending a message back to the task consumer,  the task consumer then acknowledges the task by sending `basic.ack` to the broker.

* [ Consumer ] receives task message
* [ Consumer ] delegates task message to pool

* [ PoolWorker-1] receives task message
* [ PoolWorker-1] sends ack back to the consumer (I'm working on this task now)

* [Consumer] receives ack from pool worker, and tells the broker to ack the message.

These two threads happen in parallel.  Only the broker connection that receives the task is allowed
to ack the message, and broker connections  cannot be shared between threads, so only the consumer thread is allowed to read/write to the broker socket.
